### ISSUE TITLE
Scope of MAX_EXCUTION_TIME hint is incorrect.

### ISSUE THREAD

**Issue by dveeden** [pingcap/tidb#45295]
## Bug Report

From https://dev.mysql.com/doc/refman/8.0/en/optimizer-hints.html#optimizer-hints-execution-time

"For statements with multiple SELECT keywords, such as unions or statements with subqueries, `MAX_EXECUTION_TIME` applies to the entire statement and must appear after the first `SELECT`."

### 1. Minimal reproduce step (Required)

```sql
SELECT /*+ MAX_EXECUTION_TIME(1000) */ 1 UNION ALL SELECT SLEEP(2);
SELECT /*+ MAX_EXECUTION_TIME(3000) */ 1 UNION ALL SELECT SLEEP(2);
SELECT /*+ MAX_EXECUTION_TIME(1000) */ SLEEP(2) UNION ALL SELECT 1;
SELECT /*+ MAX_EXECUTION_TIME(3000) */ SLEEP(2) UNION ALL SELECT 1;
```

### 2. What did you expect to see? (Required)

MySQL 8.0
```
mysql> SELECT /*+ MAX_EXECUTION_TIME(1000) */ 1 UNION ALL SELECT SLEEP(2);
ERROR 3024 (HY000): Query execution was interrupted, maximum statement execution time exceeded
mysql> SELECT /*+ MAX_EXECUTION_TIME(3000) */ 1 UNION ALL SELECT SLEEP(2);
+---+
| 1 |
+---+
| 1 |
| 0 |
+---+
2 rows in set (2.00 sec)

mysql> SELECT /*+ MAX_EXECUTION_TIME(1000) */ SLEEP(2) UNION ALL SELECT 1;
ERROR 3024 (HY000): Query execution was interrupted, maximum statement execution time exceeded
mysql> SELECT /*+ MAX_EXECUTION_TIME(3000) */ SLEEP(2) UNION ALL SELECT 1;
+----------+
| SLEEP(2) |
+----------+
|        0 |
|        1 |
+----------+
2 rows in set (2.00 sec)

mysql> SELECT VERSION();
+-----------+
| VERSION() |
+-----------+
| 8.0.33    |
+-----------+
1 row in set (0.00 sec)
```

With a 1000ms (1s) execution time limit the statement is aborted both if it is in the first or second part of the union.

### 3. What did you see instead (Required)

```
mysql> SELECT /*+ MAX_EXECUTION_TIME(1000) */ 1 UNION ALL SELECT SLEEP(2);
+------+
| 1    |
+------+
|    1 |
|    0 |
+------+
2 rows in set (2.00 sec)

mysql> SELECT /*+ MAX_EXECUTION_TIME(3000) */ 1 UNION ALL SELECT SLEEP(2);
+------+
| 1    |
+------+
|    1 |
|    0 |
+------+
2 rows in set (2.02 sec)

mysql> SELECT /*+ MAX_EXECUTION_TIME(1000) */ SLEEP(2) UNION ALL SELECT 1;
+----------+
| SLEEP(2) |
+----------+
|        1 |
|        0 |
+----------+
2 rows in set (2.00 sec)

mysql> SELECT /*+ MAX_EXECUTION_TIME(3000) */ SLEEP(2) UNION ALL SELECT 1;
+----------+
| SLEEP(2) |
+----------+
|        1 |
|        0 |
+----------+
2 rows in set (2.01 sec)
```

### 4. What is your TiDB version? (Required)

```
Release Version: v7.3.0-alpha-160-g90dabc0433
Edition: Community
Git Commit Hash: 90dabc0433a75db650ac683897d264ef58d63a73
Git Branch: master
UTC Build Time: 2023-07-11 07:57:57
GoVersion: go1.20.5
Race Enabled: false
Check Table Before Drop: false
Store: unistore
```

**Comment by dveeden** [comment]
Might be related to #34344 , #37004

**Comment by lcwangchao** [comment]
Seems here: https://github.com/pingcap/tidb/blob/2cf78c7be3d115541ccd4cc14fbf411b5efa93c3/util/hint/hint_processor.go#L80 does not handle the hint in union statement

**Comment by lcwangchao** [comment]
May be it's a bug for planner

**Comment by hawkingrei** [comment]
Fixed by #50070
